app/models/study.rb

```diff
@@ -1095,15 +1091,26 @@ def default_cluster_point_alpha
   # helper method to get number of unique single cells
   def set_cell_count
     cell_count = self.all_cells_array.size
-    self.update(cell_count: cell_count)
     Rails.logger.info "Setting cell count in #{self.name} to #{cell_count}"
+    self.update(cell_count: cell_count)
+    Rails.logger.info "Cell count set for #{self.name}"
   end
 
   # helper method to set the number of unique genes in this study
   def set_gene_count
-    gene_count = self.genes.pluck(:name).uniq.count
+    gene_count = self.unique_genes
     Rails.logger.info "Setting gene count in #{self.name} to #{gene_count}"
     self.update(gene_count: gene_count)
+    Rails.logger.info "Gene count set for #{self.name}"
+  end
+
+  # get all unique gene names for a study; leverage index on Gene model to improve performance
+  def unique_genes
+    genes = []
+    self.expression_matrix_files.each do |file|
+      genes += Gene.where(study_id: self.id, study_file_id: file.id).pluck(:name)
```

Comment:
I might rewrite this as `genes = Gene.where(study_id: self.id, :study_file_id.in self.expression_matrix_files.map(&:id)).pluck(:name)`  (I'm not sure I'm remembering the mongoid 'in' syntax appropriately, but you get the idea).  Just to avoid the for loop

Reply:
We actually do that in other places... good call.

Reply:
5d68378